File: chaincode/sensorchain/src/calibration.ts
/*
 * Calibration Certificate Anchor contract (production chaincode).
 * Mirrors sensorchain/contracts/calibration.py.
 */

import { createHash } from "crypto";
import { Context, Contract, Info, Returns, Transaction } from "fabric-contract-api";
import { deviceKey } from "./deviceRegistry";

const VALIDITY_SECONDS = 180 * 24 * 3600; // 180-day calibration cycle

export interface CalibrationCertificate {
  device_id: string;
  lab_id: string;
  lab_accreditation: string;
  result: string; // "pass" | "fail"
  parameters: Record<string, unknown>;
  issued_at: number;
  expires_at: number;
  certificate_hash?: string;
}

@Info({
  title: "Calibration",
  description: "Anchors lab-signed calibration certificates; auto-flags uncalibrated devices",
})
export class CalibrationContract extends Contract {
  constructor() {
    super("Calibration");
  }

  private key(ctx: Context, deviceId: string, issuedAt: number): string {
    return ctx.stub.createCompositeKey("calibration", [deviceId, issuedAt.toFixed(6).padStart(18, "0")]);
  }

  @Transaction()
  @Returns("string")
  public async AnchorCertificate(
    ctx: Context, deviceId: string, labId: string,
    labAccreditation: string, result: string, parametersJson: string,
  ): Promise<string> {
    const device = await ctx.stub.getState(deviceKey(ctx, deviceId));
    if (device.length === 0) {
      throw new Error(`cannot calibrate unregistered device ${deviceId}`);
    }
    const issuedAt = ctx.stub.getTxTimestamp().seconds.toNumber();
    const certificate: CalibrationCertificate = {
      device_id: deviceId,
      lab_id: labId,
      lab_accreditation: labAccreditation,
      result,
      parameters: parametersJson ? JSON.parse(parametersJson) : {},
      issued_at: issuedAt,
      expires_at: issuedAt + VALIDITY_SECONDS,
    };
    certificate.certificate_hash = createHash("sha256")
      .update(JSON.stringify(certificate))
      .digest("hex");
    await ctx.stub.putState(this.key(ctx, deviceId, issuedAt), Buffer.from(JSON.stringify(certificate)));
    ctx.stub.setEvent("CalibrationAnchored", Buffer.from(JSON.stringify({
      device_id: deviceId, lab_id: labId, result,
      certificate_hash: certificate.certificate_hash,
    })));
    return JSON.stringify(certificate);
  }

  @Transaction(false)
  @Returns("string")
  public async CalibrationStatus(ctx: Context, deviceId: string): Promise<string> {
    const iterator = ctx.stub.getStateByPartialCompositeKey("calibration", [deviceId]);
    let latest: CalibrationCertificate | null = null;
    for await (const kv of iterator) {
      const certificate = JSON.parse(kv.value.toString()) as CalibrationCertificate;
      if (latest === null || certificate.issued_at > latest.issued_at) {
        latest = certificate;
      }
    }
    const now = ctx.stub.getTxTimestamp().seconds.toNumber();
    if (latest === null) {
      return JSON.stringify({ device_id: deviceId, calibrated: false, reason: "never_calibrated" });
    }
    if (latest.result !== "pass") {
      return JSON.stringify({
        device_id: deviceId, calibrated: false, reason: "last_calibration_failed",
        certificate_hash: latest.certificate_hash,
      });
    }
    if (now > latest.expires_at) {
      return JSON.stringify({
        device_id: deviceId, calibrated: false, reason: "calibration_expired",
        expired_at: latest.expires_at, certificate_hash: latest.certificate_hash,
      });
    }
    return JSON.stringify({
      device_id: deviceId, calibrated: true,
      expires_at: latest.expires_at, certificate_hash: latest.certificate_hash,
    });
  }

  @Transaction()
  @Returns("string")
  public async FlagUncalibrated(ctx: Context, deviceIdsJson: string): Promise<string> {
    const deviceIds = JSON.parse(deviceIdsJson) as string[];
    const flagged: unknown[] = [];
    for (const deviceId of deviceIds) {
      const status = JSON.parse(await this.CalibrationStatus(ctx, deviceId));
      if (!status.calibrated) {
        ctx.stub.setEvent("UncalibratedDevice", Buffer.from(JSON.stringify(status)));
        flagged.push(status);
      }
    }
    return JSON.stringify(flagged);
  }
}

File: sensorchain/contracts/calibration.py
"""Calibration Certificate Anchor contract.

Certifying labs sign periodic calibration certificates; the signed
certificate hash is anchored on-chain against the device. A device
whose latest certificate has expired (or that never had one) is
auto-flagged, and its subsequent readings carry a degraded-trust
marker in the verification portal.
"""

from __future__ import annotations

import hashlib
import time

from ..ledger import Ledger
from ..merkle import canonical_json


class CalibrationContract:
    CONTRACT = "calibration"

    def __init__(self, ledger: Ledger, validity_seconds: float = 180 * 24 * 3600):
        self.ledger = ledger
        self.validity_seconds = validity_seconds  # default: 180-day calibration cycle

    @staticmethod
    def _key(device_id: str, issued_at: float) -> str:
        return f"calibration~{device_id}~{issued_at:017.6f}"

    def anchor_certificate(
        self,
        device_id: str,
        lab_id: str,
        lab_accreditation: str,
        result: str,
        issued_at: float | None = None,
        parameters: dict | None = None,
    ) -> dict:
        """Anchor a calibration certificate hash for a registered device."""
        if self.ledger.get_state(f"device~{device_id}") is None:
            raise KeyError(f"cannot calibrate unregistered device {device_id}")
        issued_at = issued_at if issued_at is not None else time.time()
        certificate = {
            "device_id": device_id,
            "lab_id": lab_id,
            "lab_accreditation": lab_accreditation,   # e.g. NABL accreditation number
            "result": result,                          # "pass" | "fail"
            "parameters": parameters or {},
            "issued_at": issued_at,
            "expires_at": issued_at + self.validity_seconds,
        }
        certificate["certificate_hash"] = hashlib.sha256(canonical_json(certificate)).hexdigest()
        tx = self.ledger.put_state(
            self.CONTRACT, "anchorCertificate", self._key(device_id, issued_at), certificate
        )
        self.ledger.emit_event("CalibrationAnchored", {
            "device_id": device_id,
            "lab_id": lab_id,
            "result": result,
            "certificate_hash": certificate["certificate_hash"],
            "tx_id": tx.tx_id,
        })
        return certificate

    def latest_certificate(self, device_id: str) -> dict | None:
        certs = [v for _, v in self.ledger.query_by_prefix(f"calibration~{device_id}~")]
        return max(certs, key=lambda c: c["issued_at"]) if certs else None

    def calibration_status(self, device_id: str, now: float | None = None) -> dict:
        """Trust status used by the anchor contract and the portal."""
        now = now if now is not None else time.time()
        cert = self.latest_certificate(device_id)
        if cert is None:
            return {"device_id": device_id, "calibrated": False, "reason": "never_calibrated"}
        if cert["result"] != "pass":
            return {"device_id": device_id, "calibrated": False, "reason": "last_calibration_failed",
                    "certificate_hash": cert["certificate_hash"]}
        if now > cert["expires_at"]:
            return {"device_id": device_id, "calibrated": False, "reason": "calibration_expired",
                    "expired_at": cert["expires_at"], "certificate_hash": cert["certificate_hash"]}
        return {"device_id": device_id, "calibrated": True,
                "expires_at": cert["expires_at"], "certificate_hash": cert["certificate_hash"]}

    def flag_uncalibrated(self, device_ids: list[str], now: float | None = None) -> list[dict]:
        """Sweep a device list; emit UncalibratedDevice events for offenders."""
        flagged = []
        for device_id in device_ids:
            status = self.calibration_status(device_id, now=now)
            if not status["calibrated"]:
                self.ledger.emit_event("UncalibratedDevice", status)
                flagged.append(status)
        return flagged
